leapsectz/leapsectz.go

```diff
@@ -88,34 +143,176 @@ func parse(r io.Reader) ([]LeapSecond, error) {
 		var nn uint32
 		err := binary.Read(r, binary.BigEndian, &nn)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 		n[i] = int(nn)
 	}
 
 	// skip uninteresting data:
-	//  tzh_timecnt (char [4])s  coded transition times a la time(2)
+	//  tzh_timecnt (char [4] or char [8] for ver 2)s  coded transition times a la time(2)
 	//  tzh_timecnt (unsigned char)s types of local time starting at above
 	//  tzh_typecnt repetitions of
 	//   one (char [4])  coded UT offset in seconds
 	//   one (unsigned char) used to set tm_isdst
 	//   one (unsigned char) that's an abbreviation list index
 	//  tzh_charcnt (char)s  '\0'-terminated zone abbreviations
-	skip := n[NTime]*5 + n[NZone]*6 + n[NChar]
-	data := make([]byte, skip)
-	if n, _ := r.Read(data); n != skip {
-		return nil, errBadData
+	var skip int
+	if version == 0 {
+		skip = n[NTime]*5 + n[NZone]*6 + n[NChar]
+	} else {
+		skip = n[NTime]*9 + n[NZone]*6 + n[NChar]
+	}
+	if n, _ := io.CopyN(ioutil.Discard, r, int64(skip)); n != int64(skip) {
+		return nil, nil, errBadData
 	}
 
-	var ret []LeapSecond
-	for i := 0; i < int(n[NLeap]); i++ {
-		l := LeapSecond{}
-		err := binary.Read(r, binary.BigEndian, &l)
-		if err != nil {
-			return nil, err
+	// calculate the amount of bytes to skip after reading leap seconds array
+	skip = n[NUTCLocal] + n[NStdWall]
+
+	if version == 0 {
+		var ret []LeapSecond
+		for i := 0; i < int(n[NLeap]); i++ {
+			l := LeapSecond{}
+			err := binary.Read(r, binary.BigEndian, &l)
+			if err != nil {
+				return nil, nil, err
+			}
+			ret = append(ret, l)
+		}
+		// we need to skip the rest of the data
+		_, _ = io.CopyN(ioutil.Discard, r, int64(skip))
+		return ret, nil, nil
+	} else {
+		var ret []LeapSecondV2
+		for i := 0; i < int(n[NLeap]); i++ {
+			l := LeapSecondV2{}
+			err := binary.Read(r, binary.BigEndian, &l)
+			if err != nil {
+				return nil, nil, err
+			}
+			ret = append(ret, l)
 		}
-		ret = append(ret, l)
+		_, _ = io.CopyN(ioutil.Discard, r, int64(skip))
+		return nil, ret, nil
+	}
+}
+
+func prepareHeader(ver byte, lsCnt int, name string) []byte {
+	const magicHeader = "TZif"
+
+	h := new(bytes.Buffer)
+
+	hdr := Header{1, 1, uint32(lsCnt), 0, 1, uint32(len(name))}
```

Comment:
I strongly suggest using named initialization, i.e. `Header{foor: 1, bar: 2}`, this is not readable.

Reply:
Made initialization named